File: camo-gen/src/texture.cpp
#include "texture.h"

#include <cstdint>

Texture::Texture(uint32_t width, uint32_t height) {
    this->width = width;
    this->height = height;
    data_buffer = new uint8_t[width * height * 4];
    if (data_buffer) isInit = true;
}

Texture::~Texture() {
    delete [] data_buffer;
}

void Texture::set_pixel(uint32_t x, uint32_t y, Color c) {
    if (isInit) {
        uint32_t offset = (y*(width*4)) + (x*4);
        data_buffer[offset + 0] = c.red;
        data_buffer[offset + 1] = c.green;
        data_buffer[offset + 2] = c.blue;
        data_buffer[offset + 3] = c.alpha;
    }
}

Color Texture::get_pixel(uint32_t x, uint32_t y) {
    Color c;
    if (isInit) {
        uint32_t offset = (y*(width*4)) + (x*4);
        c.red   = data_buffer[offset + 0];
        c.green = data_buffer[offset + 1];
        c.blue  = data_buffer[offset + 2];
        c.alpha = data_buffer[offset + 3];
    } else {
        c.red = 0;
        c.green = 0;
        c.blue = 0;
        c.alpha = 255;
    }
    return c;
}

const uint8_t *const Texture::data() {
    return (const uint8_t *const)data_buffer;
}

File: camo-gen/camogen/src/texture.rs
#[derive(Debug)]
pub struct Color {
    pub red: u8,
    pub green: u8,
    pub blue: u8,
    pub alpha: u8,
}

pub struct Texture {
    pub data_buffer: Vec<u8>,
    pub width: usize,
    pub height: usize,
}

impl Texture {
    pub fn new(width: usize, height: usize) -> Texture {
        let mut data = Vec::new();
        data.resize((width * height * 4) as usize, 0);
        Texture {
            data_buffer: data,
            width,
            height,
        }
    }

    pub fn set_pixel(&mut self, x: usize, y: usize, c: Color) {
        let mut offset = (y * (self.width * 4)) + (x * 4);
        if offset >= self.data_buffer.len() - 4 {
            offset = self.data_buffer.len() - 4;
        }
        self.data_buffer[offset] = c.red;
        self.data_buffer[offset + 1] = c.green;
        self.data_buffer[offset + 2] = c.blue;
        self.data_buffer[offset + 3] = c.alpha;
    }

    pub fn get_pixel(&self, x: usize, y: usize) -> Color {
        let mut offset = (y * (self.width * 4)) + (x * 4);
        if offset >= self.data_buffer.len() - 4 {
            offset = self.data_buffer.len() - 4;
        }
        Color {
            red: self.data_buffer[offset],
            green: self.data_buffer[offset + 1],
            blue: self.data_buffer[offset + 2],
            alpha: self.data_buffer[offset + 3],
        }
    }
}
